### PR TITLE
Add a tolerance for merge_triangulations.

### PR DESCRIPTION
**Pull request by drwells**
The default setting is not robust with respect to roundoff errors (or other minor numerical errors).

Part of the discussion around #6663: the test I added is exactly the same grid as I showed in the mailing list. Without this patch we get a grid that looks like (you may have to zoom in to see the issue)

grid

because the vertices from the inner grid are not combined with the vertices of the background Cartesian grid.

### CODE REVIEW

File: include/deal.II/grid/grid_generator.h [hunk 1]
```diff
@@ -996,7 +996,8 @@ namespace GridGenerator
   void
   merge_triangulations(const Triangulation<dim, spacedim> &triangulation_1,
                        const Triangulation<dim, spacedim> &triangulation_2,
-                       Triangulation<dim, spacedim> &      result);
+                       Triangulation<dim, spacedim> &      result,
+                       const double duplicated_vertex_tolerance = 1.0e-12);
```
Comment (bangerth) [inline include/deal.II/grid/grid_generator.h]: can you document this argument?

File: source/grid/grid_generator.cc [hunk 2]
```diff
@@ -3784,7 +3784,8 @@ namespace GridGenerator
   void
   merge_triangulations(const Triangulation<dim, spacedim> &triangulation_1,
                        const Triangulation<dim, spacedim> &triangulation_2,
-                       Triangulation<dim, spacedim> &      result)
+                       Triangulation<dim, spacedim> &      result,
+                       const double duplicated_vertex_tolerance)
```
Comment (bangerth) [inline source/grid/grid_generator.cc]: align it if you can and clang-format doesn't undo it

File: source/grid/grid_generator.cc [hunk 3]
```diff
@@ -3831,8 +3832,11 @@ namespace GridGenerator
     // necessary and create the triangulation
     SubCellData               subcell_data;
     std::vector<unsigned int> considered_vertices;
-    GridTools::delete_duplicated_vertices(
-      vertices, cells, subcell_data, considered_vertices);
+    GridTools::delete_duplicated_vertices(vertices,
+                                          cells,
+                                          subcell_data,
+                                          considered_vertices,
+                                          duplicated_vertex_tolerance);
```
Comment (bangerth) [inline source/grid/grid_generator.cc]: ah, that was easy :-)

File: source/grid/grid_generator.cc (hunk 2 above)
Comment (drwells) [reply source/grid/grid_generator.cc]: clang-format always puts it this way :(

File: include/deal.II/grid/grid_generator.h (hunk 1 above)
Comment (drwells) [reply include/deal.II/grid/grid_generator.h]: Done.

**Comment by drwells** [comment]
Wow, clang-format really didn't like the `@code` block. Let me fix that.

File: include/deal.II/grid/grid_generator.h (hunk 1 above)
Comment (drwells) [reply include/deal.II/grid/grid_generator.h]: Done.

**Comment by drwells** [comment]
This is now ready.

**Comment by bangerth** [comment]
OK to merge once the tester is happy.

/run-tests

**Pull request merged**